(beta) Supports Sub in Slack

srajiang opened the pull request:
###  Summary

👋 This PR adds initial support for Subscribe in Slack intended for beta release.

#### Updates
* Adds a `app.subscription` method to `App.ts` which accepts user provided middleware options needed for handling `notification_subscriptions_*` events. 
* Adds a Subscription class implementation which registers middleware.
* Adds relevant types and adds tests for the functions of the above class.

#### Testing
- Here is a [working sample app](https://github.com/srajiang/slack-subscribe-sample) using Bolt JS. 
   - The sample uses `client.apps.notifications.subscriptions.update` method. This calls _an in progress endpoint_ which returns an error currently (confirmed). For beta release, I have included this support to call this endpoint, but this will need to be confirmed with server teams when they have completed the support for the endpoint or removed it entirely. 

#### Todo 
- [x] Add a sample application
- [x] Resolve blocking types-tests [EDIT]: Resolved by reverting a commit introduce likely due to some funky rebasing. I have noted where these places are and confirm that they matched with main. 

### Requirements (place an `x` in each `[ ]`)
* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/bolt/blob/main/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).

srajiang commented on src/App.ts:
```diff
@@ -524,6 +527,23 @@ export default class App {
+   *
+   * @param listeners middleware that process and react to subscription payloads
+   */
+  public subscription(options: SubscriptionOptions): this {
```
Note: Originally the design called for this method to accept a first parameter for a `custom_id` that would correspond to specific resource domains (set up during Slack App configuration). The `custom_id` would have had to be sent from Slack backend. Since this feature will not be implemented at this time, adding this custom_id field later would be a breaking change.

srajiang commented on src/App.ts:
```diff
@@ -524,6 +527,23 @@ export default class App {
+   */
+  public subscription(options: SubscriptionOptions): this {
+    /*
+     * TODO If it becomes possible for users to set a custom unique resource id,
```
Breadcrumb for further expansion to include `custom_id`

srajiang commented on src/Subscription.spec.ts:
```diff
@@ -0,0 +1,230 @@
+  });
+
+  describe('matchesConstraints()', () => {
+    // TODO: Implement
```
Added a test stub here when matchesConstraints is able filter events that match custom_id constraint

filmaj commented:
Haven't done a full review here but just looking at the type failures. I just had time to dig into the first failure:

```
  types-tests/options.test-d.ts:26:32
  ✖   26:32  Argument of type SlackOptions is not assignable to parameter of type BlockSuggestion.
  Type InteractiveMessageSuggestion is missing the following properties from type BlockSuggestion: block_id, action_id, api_app_id, container
```

Looking at this test, I'm wondering if the type of the `options` argument should be `BlockSuggestion`? It seems like `SlackOptions` is the overarching supertype in this situation (via [`payload` being `OptionsPayloadFromType`](https://github.com/slackapi/bolt-js/blob/main/src/types/options/index.ts#L10) which seems to eventually build off [`SlackOptions`](https://github.com/slackapi/bolt-js/blob/main/src/types/options/index.ts#L32)). I am by no means a TypeScript expert, and digging into this hurt my brain, but changing the expected type [in the test](https://github.com/slackapi/bolt-js/blob/main/types-tests/options.test-d.ts#L26) from `BlockSuggestion` to `SlackOptions` at least addressed the first error.

srajiang commented:
@filmaj - Thanks for looking into this. I also started along this thinking, but stopped short of trying to resolve those Typescript compiler complaints because I'm very unfamiliar with what the type-tests are doing and also find the errors so difficult to parse. 

Some things I've tried - 
* rebased feat-hermes off the latest main and rebased this development branch off feat-hermes. 
* I've also been trying different combinations of @slack/web-api and @slack/types dependencies to see if they can make the complaints go away, but haven't had much luck

stevengill reviewed (approved):
Looks great

misscoded commented on src/types/subscription/index.ts:
```diff
@@ -0,0 +1,100 @@
+export type SubscriptionOnDeletedMiddleware = Middleware<SubscriptionMiddlewareArgs>;
+
+export interface SubConfigureFn {
+  (params: View): Promise<ViewsOpenResponse>,
```
```suggestion
  (params: View): Promise<ViewsOpenResponse>;
```

misscoded commented on src/types/subscription/index.ts:
```diff
@@ -0,0 +1,100 @@
+  (params: View): Promise<ViewsOpenResponse>,
+}
+export interface SubSayFn {
+  (params: SayArguments): Promise<ChatPostMessageResponse>,
```
```suggestion
  (params: SayArguments): Promise<ChatPostMessageResponse>;
```

srajiang commented on src/App.ts:
```diff
@@ -12,7 +12,6 @@ import {
   matchConstraints,
   onlyCommands,
   matchCommandName,
-  matchFunctionCallback,
```
Removed due to reverted commit included in history after bad rebase. The changed version should match `main`. To verify: [L15](https://github.com/slackapi/bolt-js/blob/237a503d1e19a8e40ca9ed130dacab1be9752d53/src/App.ts#L15)

srajiang commented on src/types/events/base-events.ts:
```diff
@@ -898,28 +897,3 @@ export interface WorkflowStepExecuteEvent {
 // NOTE: `user_resourced_denied`, `user_resource_granted`, `user_resourced_removed` are left out because they are
 // deprecated as part of the Workspace Apps Developer Preview
-
-export interface FunctionExecutedEvent {
```
Removed due to reverted commit included in history after bad rebase. The changed version should match `main`. To verify: [L899](https://github.com/slackapi/bolt-js/blob/237a503d1e19a8e40ca9ed130dacab1be9752d53/src/types/events/base-events.ts#L899)

srajiang commented on src/types/events/base-events.ts:
```diff
@@ -75,8 +75,7 @@ export type SlackEvent =
   | WorkflowUnpublishedEvent
   | WorkflowStepDeletedEvent
-  | WorkflowStepExecuteEvent
-  | FunctionExecutedEvent;
```
Removed due to reverted commit included in history after bad rebase. The changed version should match `main`. To verify: [L78](https://github.com/slackapi/bolt-js/blob/237a503d1e19a8e40ca9ed130dacab1be9752d53/src/types/events/base-events.ts#L78)

srajiang commented on src/middleware/builtin.ts:
```diff
@@ -251,22 +251,6 @@ function matchesPattern(pattern: string | RegExp, candidate: string): boolean {
   return pattern.test(candidate);
 }
 
-/**
```
Removed due to reverted commit included in history after bad rebase. The changed version should match `main`. To verify: [L254](https://github.com/slackapi/bolt-js/blob/237a503d1e19a8e40ca9ed130dacab1be9752d53/src/middleware/builtin.ts#L254)

srajiang commented on src/App.ts:
```diff
@@ -692,12 +711,7 @@ export default class App {
     this.listeners.push([onlyCommands, matchCommandName(commandName), ...listeners] as Middleware<AnyMiddlewareArgs>[]);
   }
 
-  // matches a Slack function callback specifically
```
Removed due to reverted commit included in history after bad rebase. The changed version should match main. To verify: [L694](https://github.com/slackapi/bolt-js/blob/237a503d1e19a8e40ca9ed130dacab1be9752d53/src/App.ts#L694)
